File: backend/Timesheets.Api/Auth/Endpoints/GetCurrentUserPermissions.cs
using Timesheets.Api.Auth;
using Timesheets.Api.Common;
using Timesheets.Api.Common.Extensions;

namespace Timesheets.Api.Auth.Endpoints;

public sealed class GetCurrentUserPermissions : IEndpoint
{
    public static void Map(IEndpointRouteBuilder app) =>
        app.MapGet("/currentUserPermissions", Handle)
           .WithSummary("Get Currently Authenticated User Permissions")
           .AddEndpointFilter(RequireAuthenticatedFilter.Instance)
           .AddEndpointFilter<EnsureCurrentUserLoadedFilter>();

    public sealed record Response(
        UserRole Role,
        IReadOnlyList<Guid> ProjectManagerOf,
        IReadOnlyList<Guid> ContractManagerOf,
        IReadOnlyList<Guid> EmployeeOnContractIds,
        IReadOnlyList<Guid> VisibleProjectIds,
        IReadOnlyList<Guid> VisibleContractIds);

    private static Task<IResult> Handle(ICurrentUser user) =>
        Task.FromResult<IResult>(Results.Ok(new Response(
            Role: user.Role,
            ProjectManagerOf: user.ProjectManagerOf,
            ContractManagerOf: user.ContractManagerOf,
            EmployeeOnContractIds: user.EmployeeOnContractIds,
            VisibleProjectIds: user.VisibleProjectIds,
            VisibleContractIds: user.VisibleContractIds)));

    private sealed class RequireAuthenticatedFilter : IEndpointFilter
    {
        public static readonly RequireAuthenticatedFilter Instance = new();

        public ValueTask<object?> InvokeAsync(EndpointFilterInvocationContext context, EndpointFilterDelegate next) =>
            context.HttpContext.User.IsAuthenticated()
                ? next(context)
                : ValueTask.FromResult<object?>(Results.Unauthorized());
    }
}

File: frontend/src/auth/api/getCurrentUserPermissions.ts
import { BaseUrl, customFetch, withDelay } from "@/constants/api";
import type { UserRole } from "../userRole";

export interface CurrentUserPermissions {
  role: UserRole;
  projectManagerOf: string[];
  contractManagerOf: string[];
  employeeOnContractIds: string[];
  visibleProjectIds: string[];
  visibleContractIds: string[];
}

interface GetCurrentUserPermissionsResponse {
  role: UserRole;
  projectManagerOf: string[];
  contractManagerOf: string[];
  employeeOnContractIds: string[];
  visibleProjectIds: string[];
  visibleContractIds: string[];
}

export const getCurrentUserPermissions = () =>
  withDelay("fast", () =>
    customFetch<GetCurrentUserPermissionsResponse>(`${BaseUrl}/auth/currentUserPermissions`, {
      credentials: "include",
    }),
  );
